
<h1 id="Refinitive-Test">Refinitive Test<a class="anchor-link" href="#Refinitive-Test">¶</a></h1>



<h2 id="Question-1">Question 1<a class="anchor-link" href="#Question-1">¶</a></h2>



<p>Examine the dataset in sheet 'Question1_Data' and answer the following questions:<br/>
1.1 How many employees are currently employed (Inactive = 0)? Enter the formula in column P<br/>
1.2 How many employees are actively employed in the IT Department in Cape Town? Enter the formula in column P<br/>
1.3 What is the total number of leave days for all active employees in Cape Town? Enter the formula in column P</p>



<p>Begin by reading in the excel book and creating data frames in pandas</p>


In [21]:

from pathlib import Path
import pandas as pd
import numpy as np


In [22]:

p = Path.home() / 'refinitiv-test'
xls = pd.ExcelFile(p / 'ExcelTest1.2.xlsx')
df_list = [pd.read_excel(xls, sheet) for sheet in xls.sheet_names]
_, data1, _, data2, _, data3, _ = df_list




<p>The first dataset <code>data1</code> has empty columns so we drop them</p>


In [23]:

data1.dropna(axis=1, how='all', inplace=True)
print(data1.head())



   EmployeeID DepartmentID       Office  InActive  Employee ID  LeaveDays
0           1           HR  East London         0            1          7
1           2           HR     Pretoria         0            2         13
2           3  Procurement    Cape Town         1            3         12
3           4  Procurement       Durban         1            4          7
4           5          R&D    Cape Town         0            5          6



<h3 id="1.1">1.1<a class="anchor-link" href="#1.1">¶</a></h3><p><strong>How many employees are currently employed (Inactive = 0)? Enter the formula in column P</strong></p>


In [24]:

# Question 1.1
data1.InActive.value_counts()
    



InActive
1    1112
0    1097
Name: count, dtype: int64


<p>We see that there are 1097 active employees</p>



<h3 id="1.2">1.2<a class="anchor-link" href="#1.2">¶</a></h3><p><strong>How many employees are actively employed in the IT Department in Cape Town? Enter the formula in column P</strong></p>



<p>We take the rows for the IT department that have an <code>InActive</code> code of <code>0</code></p>


In [25]:

# Question 1.2
len(data1[(data1.DepartmentID == 'IT') & (data1.InActive == 0)])



118


<p>There are 118 people employed in the IT department in Cape Town</p>



<h3 id="1.3">1.3<a class="anchor-link" href="#1.3">¶</a></h3><p><strong>What is the total number of leave days for all active employees in Cape Town? Enter the formula in column P</strong></p>



<p>Start by getting all the active employees in Cape Town</p>


In [26]:

cape_town = data1[(data1.Office == 'Cape Town') & (data1.InActive==0)]
print(cape_town.head())



    EmployeeID DepartmentID     Office  InActive  Employee ID  LeaveDays
4            5          R&D  Cape Town         0            5          6
6            7      Payroll  Cape Town         0            7         10
14          15          R&D  Cape Town         0           15         12
31          32      Payroll  Cape Town         0           32          7
32          33           HR  Cape Town         0           33         14



<p>Then sum up the leave days</p>


In [27]:

cape_town.LeaveDays.sum()



2379


<p>There are a total of 2379 leave days</p>



<h2 id="Question-2">Question 2<a class="anchor-link" href="#Question-2">¶</a></h2>



<p>Examine the dataset in sheet 'Question2_Data' and answer the following questions:<br/>
2.1 How many distinct UIDs contain the Source URL 'www.abc.com'?<br/>
2.2 How many distinct UIDs contain duplicate Source URLs?<br/>
2.3 How many distinct Source URLs does UID 2989625 have?</p>



<h3 id="2.1">2.1<a class="anchor-link" href="#2.1">¶</a></h3><p><strong>How many distinct UIDs contain the Source URL 'www.abc.com'?</strong></p>



<p>First we look for the distinct UIDs</p>


In [28]:

data2_dedup = data2.drop_duplicates(subset='UID')
print('There are ', len(data2.UID), 'UIDs in the data')
print('There are ', data2.UID.nunique(), 'distinct UIDs in the data')



There are  1329 UIDs in the data
There are  1219 distinct UIDs in the data



<p>Next, search for the string www.abc.com in the Source URL string</p>


In [29]:

data2_dedup['Source URLs'].str.contains('www.abc.com').sum()



568


<p>There are 568 distinct UIDs that contain the source url 'www.abc.com'</p>



<h3 id="2.2">2.2<a class="anchor-link" href="#2.2">¶</a></h3><p><strong>How many distinct UIDs contain duplicate Source URLs?</strong></p>



<p>Using our de-duplicated data on UID, check for duplicates in the Source URL</p>


In [30]:

data2_dedup['Source URLs'].duplicated().sum()



1162


<p>There are 1,162 duplicates in the Source URLs</p>



<h3 id="2.3">2.3<a class="anchor-link" href="#2.3">¶</a></h3><p><strong>How many distinct Source URLs does UID 2989625 have?</strong></p>



<p>We check for rows where UID is 2989625 and then count the unique Source URLs for those rows</p>


In [31]:

data2[data2.UID == 2989625]['Source URLs'].nunique()



1


<p>There is only one unique source URL for UID 2989625</p>



<h3 id="2.4">2.4<a class="anchor-link" href="#2.4">¶</a></h3><p><strong>Provide the count of distinct citizens for each of the Countries in column O:</strong></p>



<p>We start by breaking the <code>Citizenship</code> column into multiple columns to account for multiple nationalities</p>


In [32]:

citizen = data2_dedup.CITIZENSHIP.str.split(';', expand=True)
citizen.columns = ['Citizenship1', 'Citizenship2', 'Citizenship3']
print(citizen.head())    



  Citizenship1 Citizenship2 Citizenship3
0          USA         None         None
1          USA   MOZAMBIQUE         None
2          USA         None         None
3          USA       CYPRUS         None
4          USA         None         None



<p>Then we get the counts for each column</p>


In [33]:

citizen.Citizenship1.value_counts()



Citizenship1
USA    1219
Name: count, dtype: int64

In [34]:

citizen.Citizenship2.value_counts()



Citizenship2
RUSSIAN FEDERATION    47
CYPRUS                33
ZAMBIA                31
MOZAMBIQUE            26
CHINA                 24
FRANCE                18
Name: count, dtype: int64

In [35]:

citizen.Citizenship3.value_counts()



Citizenship3
PARAGUAY    18
Name: count, dtype: int64


<h3 id="2.5">2.5<a class="anchor-link" href="#2.5">¶</a></h3><p><strong>Provide the count of individuals born between 1960 and 1970</strong></p>



<p>Since there are multiple dates of birth for a person, we start by separating them</p>


In [36]:

dob = data2_dedup.DOB.str.split(';', expand=True).fillna(np.nan)
dob.columns = [f'dob{i}' for i in range(dob.shape[1])]
print(dob.head())



         dob0        dob1 dob2 dob3 dob4 dob5 dob6 dob7
0  1968/00/00         NaN  NaN  NaN  NaN  NaN  NaN  NaN
1         NaN         NaN  NaN  NaN  NaN  NaN  NaN  NaN
2  1955/03/01  1953/00/00  NaN  NaN  NaN  NaN  NaN  NaN
3         NaN         NaN  NaN  NaN  NaN  NaN  NaN  NaN
4         NaN         NaN  NaN  NaN  NaN  NaN  NaN  NaN



<p>Since we are only interested in the year, we take the first part before the '/' of every string of the form 'yyyy/mm/dd'. We define a function that will extract the year portion of the string</p>


In [37]:

def date_strip(x):
    return  pd.DataFrame(x.astype(str).str.split('/').tolist()).iloc[:,0].astype(float)




<p>Applying the <code>date_strip</code> function to every column gives us</p>


In [38]:

years = dob.apply(date_strip)
print(years.head())



     dob0    dob1  dob2  dob3  dob4  dob5  dob6  dob7
0  1968.0     NaN   NaN   NaN   NaN   NaN   NaN   NaN
1     NaN     NaN   NaN   NaN   NaN   NaN   NaN   NaN
2  1955.0  1953.0   NaN   NaN   NaN   NaN   NaN   NaN
3     NaN     NaN   NaN   NaN   NaN   NaN   NaN   NaN
4     NaN     NaN   NaN   NaN   NaN   NaN   NaN   NaN



<p>Then we count the number of years of birth that are between 1960 and 1970 inclusive, being careful to check for people with multiple years that fall in the range to avoid double counting.</p>
<p>The function defined by <code>lambda</code> will count the number of birth years that are between 1960 and 1970</p>


In [39]:

years.apply(lambda x: ((x >= 1960) & (x <= 1970)).sum(), axis=1).value_counts()   



0    1176
1      37
2       6
Name: count, dtype: int64


<p>We can see that there are 37 people born between 1960 and 1970. The additional 6 are for people with different birthdates recorded but the same birth year</p>



<h2 id="Question-3">Question 3<a class="anchor-link" href="#Question-3">¶</a></h2><p>Examine the dataset in sheet 'Question3_Data' and answer the following questions:<br/>
3.1 For each of the TEJ_Code values, extract the number between underscore characters and list them in column C. What is the sum of all the extracted numbers?<br/>
3.2 Replace the extracted number in the original TEJ_Code with the corresponding KW in Column K. Place the new TEJ_Code in a new column called 'TEJ_Code_KW'</p>



<h3 id="3.1">3.1<a class="anchor-link" href="#3.1">¶</a></h3><p><strong>For each of the TEJ_Code values, extract the number between underscore characters and list them in column C. What is the sum of all the extracted numbers?</strong></p>



<p>There are columns in <code>data3</code> that are empty so we drop them</p>


In [40]:

data3.dropna(axis=1, how='all', inplace=True)




<p>Next, we extract the number part of the TEJ codes</p>


In [41]:

tej_codes = data3.TEJ_Code.str.split('_', expand=True)
print(tej_codes.head())



     0   1        2
0  TEJ  64     DGHM
1  TEJ  75    WESGV
2  TEJ  61    SATRU
3  TEJ  35     AERH
4  TEJ  51  AQEWRGT



<p>Then sum up the second column</p>


In [42]:

tej_codes.iloc[:,1].astype(int).sum()



49465


<p>The sum of the numbers is 49465</p>



<h3 id="3.2">3.2<a class="anchor-link" href="#3.2">¶</a></h3><p><strong>Replace the extracted number in the original TEJ_Code with the corresponding KW in Column K. Place the new TEJ_Code in a new column called 'TEJ_Code_KW'</strong></p>



<p>Using the <code>tej_codes</code> subset from part 3.1, we rename the columns from 0, 1, 2 to 'TEJ', 'ID', 'LAST'. We also cast the ID column as a float, so that it can be used to merge with <code>data3</code></p>


In [43]:

tej_codes.columns = ['TEJ', 'ID', 'LAST']
tej_codes.ID = tej_codes.ID.astype(float)  
print(tej_codes.head())



   TEJ    ID     LAST
0  TEJ  64.0     DGHM
1  TEJ  75.0    WESGV
2  TEJ  61.0    SATRU
3  TEJ  35.0     AERH
4  TEJ  51.0  AQEWRGT



<p>Next, we match the integer parts with the ID column. After matching, we use the corresponding <code>KW</code> to create the new <code>TEJ_Code_KW</code> column.</p>


In [44]:

merged = tej_codes.merge(data3, on='ID')
print(merged.head())



   TEJ    ID     LAST     UID       TEJ_Code   KW
0  TEJ  64.0     DGHM  289079    TEJ_34_DGHM  ghi
1  TEJ  75.0    WESGV  237266   TEJ_48_AERGH  stu
2  TEJ  61.0    SATRU   26111   TEJ_72_ADFRH  ijk
3  TEJ  35.0     AERH  260015  TEJ_60_ZDGTJH  lmn
4  TEJ  51.0  AQEWRGT   59990    TEJ_75_AE5Y  vwx



<p>Finally, <code>TEJ_Code_KW</code> is created by concatenating the string <code>TEJ</code>, the <code>KW</code> and the last 4 characters.</p>


In [45]:

merged['TEJ_Code_KW'] = merged['TEJ'] + '_' + merged['KW'] + '_' + merged['LAST']
print(merged.head())
print(merged.tail())



   TEJ    ID     LAST     UID       TEJ_Code   KW      TEJ_Code_KW
0  TEJ  64.0     DGHM  289079    TEJ_34_DGHM  ghi     TEJ_ghi_DGHM
1  TEJ  75.0    WESGV  237266   TEJ_48_AERGH  stu    TEJ_stu_WESGV
2  TEJ  61.0    SATRU   26111   TEJ_72_ADFRH  ijk    TEJ_ijk_SATRU
3  TEJ  35.0     AERH  260015  TEJ_60_ZDGTJH  lmn     TEJ_lmn_AERH
4  TEJ  51.0  AQEWRGT   59990    TEJ_75_AE5Y  vwx  TEJ_vwx_AQEWRGT
     TEJ    ID     LAST     UID          TEJ_Code   KW      TEJ_Code_KW
994  TEJ  73.0  ASFDREH   49161      TEJ_29_ASFDG  pqr  TEJ_pqr_ASFDREH
995  TEJ  75.0    DCGYJ  237266      TEJ_48_AERGH  stu    TEJ_stu_DCGYJ
996  TEJ  39.0     SDFH   50976  TEJ_80_DFGSAEFDG  vwx     TEJ_vwx_SDFH
997  TEJ  32.0    ADFRH   84768       TEJ_39_AE5Y  cde    TEJ_cde_ADFRH
998  TEJ  77.0     SADF  156696      TEJ_66_ASRTU  ghi     TEJ_ghi_SADF
